Feature/optimize a2dp avdtp

MarkWangChinese opened the pull request:
- When the cmd packet is invalid (buf->len is not enough), reply with reject.
- Split the packet handler to cmd_handler and rsp_handler, and optimize the handler array to be function pointer array.
- If receiving the unsupported cmd (signal identifier), send reject or discard it.
- Optimize `bt_a2dp_ep_info`'s field `struct bt_avdtp_sep_info sep_info` as `struct bt_avdtp_sep_info *sep_info`, see the explain in the commit.

MarkWangChinese committed:
bluetooth: a2dp: check the invalid avdtp packet

When the packet is invalid, reply the right error code or call
back error code to upper layer.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
bluetooth: avdtp: split packet handler as cmd_handler and rsp_handler

use two handler array to process cmd and response separately,
make the functions shorter and codes more readable.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
bluetooth: avdtp: handle the unsupported signal identifier

reject the cmd if the signal identifier is not supported

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
bluetooth: a2dp: optimize bt_a2dp_ep_info

bt_a2dp_ep_info is used in bt_a2dp_discover, and the max
discovered endpoints' info are already given in seps_info
of bt_a2dp_discover_param, so use pointer to point to the
seps_info of bt_a2dp_discover_param.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

jhedberg commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -260,71 +260,70 @@ static void avdtp_set_status(struct bt_avdtp_req *req, struct net_buf *buf, uint
+{
+	struct bt_avdtp_req *req = session->req;
+
+	if (req == NULL) {
+		return;
+	}
+	k_work_cancel_delayable(&session->timeout_work);
+	if (msg_type == BT_AVDTP_ACCEPT) {
+		DISCOVER_REQ(req)->buf = buf;
+	} else {
+		DISCOVER_REQ(req)->buf = NULL;
+	}
+	avdtp_set_status(req, buf, msg_type);
+	bt_avdtp_clear_req(session);
+	if (req->func != NULL) {
+		req->func(req);
+	}
```
This could use a bit more vertical whitespace to improve readability. E.g. add empty lines after `}` and before `if (...`

jhedberg commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -340,602 +339,594 @@ static struct bt_avdtp_sep *avdtp_get_sep(uint8_t stream_endpoint_id)
+	}
 
-		if (req == NULL) {
+	ret = bt_l2cap_chan_send(&session->br_chan.chan, rsp_buf);
+	if (ret) {
+		net_buf_unref(rsp_buf);
+		LOG_ERR("Error:L2CAP send fail - result = %d", ret);
+	}
+	if (!err && !ret) {
+		bt_avdtp_set_state(sep, AVDTP_IDLE);
+	}
+	avdtp_sep_unlock(sep);
```
Ok, I see that this "monolithic blocks" style is quite pervasive, so you might want to do a separate cleanup commit to fix it.

jhedberg commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -340,602 +339,594 @@ static struct bt_avdtp_sep *avdtp_get_sep(uint8_t stream_endpoint_id)
+		net_buf_unref(rsp_buf);
+		LOG_ERR("Error:L2CAP send fail - result = %d", ret);
+	}
+	if (!err && !ret) {
```
`else if (!err) {` perhaps? You should also fix the variable naming & usage here, since it's super confusing to have both `err` and `ret`. Is it possible to remove one of them? If not, you need to use names which are more descriptive.

jhedberg commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -340,602 +339,594 @@ static struct bt_avdtp_sep *avdtp_get_sep(uint8_t stream_endpoint_id)
+{
+	struct bt_avdtp_req *req = session->req;
+
+	if (req == NULL) {
+		return;
+	}
+	k_work_cancel_delayable(&session->timeout_work);
+	avdtp_set_status(req, buf, msg_type);
+	if (msg_type == BT_AVDTP_ACCEPT && req->status == BT_AVDTP_SUCCESS) {
```
Same thing here with the lack of vertical white space. Simple way to improve the readability is to have an empty line at least after every `}` and before `if (...) {`

jhedberg commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -340,602 +339,594 @@ static struct bt_avdtp_sep *avdtp_get_sep(uint8_t stream_endpoint_id)
+		bt_avdtp_set_state_lock(CTRL_REQ(req)->sep, AVDTP_OPEN);
+	} else if (msg_type == BT_AVDTP_REJECT) {
+		avdtp_handle_reject(buf, req);
+	}
+
+	if (req->status == BT_AVDTP_SUCCESS) {
```
Here you're doing it right - I just don't understand why you're not consistent about it.

jhedberg commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -340,602 +339,594 @@ static struct bt_avdtp_sep *avdtp_get_sep(uint8_t stream_endpoint_id)
+		net_buf_unref(rsp_buf);
+		LOG_ERR("Error:L2CAP send fail - result = %d", ret);
+	}
+	if (!err && !ret) {
```
Same thing here. There's a general convention in Zephyr to use either `err` or `ret` for `int` style generic error variable. The convention in Bluetooth code has been to use `err`. However, it gets really confusing when you have both - please don't do that. Note: I can see that this problem exists even before this PR, so fixing it doesn't necessarily have to be done in this commit/PR as long as it get fixed ASAP.

MarkWangChinese replied:
Hi @jhedberg  I add one commit to address all the blank line issue, please help to review. One exception that I do is: if the code `if (xx) {`'s condition use the previous simple code line's result, I put the previous simple codes lines near with the `if (xx) {`. like: https://github.com/zephyrproject-rtos/zephyr/pull/83375/files#diff-aebc3083e23e5c1f24f55bc6bf88e71ffa3ddc8dbae92db402bd4fa8402aa461L585-L587
This is my understand. please help to check whether it is right.

For `err` and `ret`, I plan to change `err` as `cmd_err` because it is result of processing the cmd, change `ret` as `l2cap_ret` because it is the l2cap send api return value in next pr after this pr is merged.

MarkWangChinese committed:
bluetooth: a2dp: improve the readability

add more blank lines before or after the code block {}

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
bluetooth: a2dp: clean few unused codes lines

memset already clean the `buf`, don't need to set as NULL.
`avdtp_set_status` already set `req->status` based on `msg_type`,
don't need to check `msg_type` again.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -497,6 +497,8 @@ static int bt_a2dp_get_sep_caps(struct bt_a2dp *a2dp)
+			memset(&a2dp->get_capabilities_param, 0U,
+			       sizeof(a2dp->get_capabilities_param));
 			a2dp->get_capabilities_param.req.func = bt_a2dp_get_capabilities_cb;
 			a2dp->get_capabilities_param.buf = NULL;
```
updated

MarkWangChinese commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -618,14 +622,14 @@ static void avdtp_open_handler(struct bt_avdtp *session, struct net_buf *buf, ui
 		k_work_cancel_delayable(&session->timeout_work);
 		avdtp_set_status(req, buf, msg_type);
-		if (msg_type == BT_AVDTP_ACCEPT) {
+		if (msg_type == BT_AVDTP_ACCEPT && req->status == BT_AVDTP_SUCCESS) {
```
updated

MarkWangChinese commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -765,7 +779,7 @@ static void avdtp_close_handler(struct bt_avdtp *session, struct net_buf *buf, u
 		k_work_cancel_delayable(&session->timeout_work);
 		avdtp_set_status(req, buf, msg_type);
-		if (msg_type == BT_AVDTP_ACCEPT) {
+		if (msg_type == BT_AVDTP_ACCEPT && req->status == BT_AVDTP_SUCCESS) {
```
updated

MarkWangChinese commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -580,6 +582,7 @@ int bt_a2dp_discover(struct bt_a2dp *a2dp, struct bt_a2dp_discover_param *param)
+	memset(&a2dp->discover_cb_param, 0U, sizeof(a2dp->discover_cb_param));
 	a2dp->discover_cb_param = param;
 	a2dp->discover_param.req.func = bt_a2dp_discover_cb;
 	a2dp->discover_param.buf = NULL;
```
updated

jhedberg commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -260,71 +265,75 @@ static void avdtp_set_status(struct bt_avdtp_req *req, struct net_buf *buf, uint
+
+	k_work_cancel_delayable(&session->timeout_work);
+
+	if (msg_type == BT_AVDTP_ACCEPT) {
+		DISCOVER_REQ(req)->buf = buf;
+	} else {
+		DISCOVER_REQ(req)->buf = NULL;
+	}
+
+	avdtp_set_status(req, buf, msg_type);
```
This looks buggy to me from a net_buf reference counting perspective. In the first branch (AVDTP_ACCEPT) you seem to moving the reference to `DISCOVER_REQ(req)->buf` whereas in the `else` branch the reference stays as-is. Yet, there's no difference in later treatment of the `buf` pointer. If you really intend to move the reference, it'd be best to set the original pointer (`buf`) immediately to NULL, so that it can't be used anymore. If you don't intend to move the reference, you should be using `net_buf_ref()` as part of the assignment to `DISCOVER_REQ(req)->buf`.

MarkWangChinese replied:
`DISCOVER_REQ(req)->buf` is set to be used by the upper (a2dp) layer, no `buf` needs to be passed to a2dp and `DISCOVER_REQ(req)->buf` is `NULL` if fail. `avdtp_set_status` set the result to be used by the upper (a2dp) layer too. So, the buf is always valid when calling `avdtp_set_status`.
Anyway your comment is right from reading only this part of codes. I will change it as the follow codes.
I have one query, is the `net_buf_ref` needed from your viewpoint? The `DISCOVER_REQ(req)->buf` should be only used in the `req->fun(req)` callback, `unref` is needed after callback. Without `ref` and `unref`, we can reduce codes. 

Note: If the upper layer (a2dp) want to keep the buf, a2dp need to ref the `DISCOVER_REQ(req)->buf` and keep it from my understand.
```
	avdtp_set_status(req, buf, msg_type);

	if (msg_type == BT_AVDTP_ACCEPT) {
		DISCOVER_REQ(req)->buf = buf;//net_buf_ref(buf);
	} else {
		DISCOVER_REQ(req)->buf = NULL;
		//buf = NULL;
	}

	bt_avdtp_clear_req(session);

	if (req->func != NULL) {
		req->func(req);
	}

	//if (buf) {
		//net_buf_unref(buf);
	//}
```

jhedberg replied:
> I have one query, is the `net_buf_ref` needed from your viewpoint?

It depends what you want to do. If you want to retain the original ref, then you need to use ref() as part of copying the pointer. However, if you *don't* need the original ref, then you can just move it to the req, however then the right thing to do would be to set the original pointer to NULL right after the move. In some ways it'd be nice if we had this helper function in the net_buf API, but until then you just need to do it yourself:

```c
struct net_buf *net_buf_move(struct net_buf **buf)
{
	struct net_buf *moved;

	moved = *buf;
	*buf = NULL;

	return moved;
}
```

and then your code would be doing:

```c
DISCOVER_REQ(req)->buf = net_buf_move(&buf);
```

> Without ref and unref, we can reduce codes.

Reference counted objects, and doing reference counting correctly, is a really common source of programming errors. Having a clear concept of who owns which reference, when references are moved, and following consistent patterns in code to reflect these things, is the best way to avoid such bugs.

MarkWangChinese replied:
Thank very mchu for the explanation. I updated it to use the `move`.

MarkWangChinese committed:
bluetooth: avdtp: move the buf as callback parameter

if saving buf to another variable, usually the ref/unref
needs to be used. change to pass the buf as one parameter
of callback

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

jhedberg commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -420,13 +422,16 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+		/* move the ref of buf to GET_CAP_REQ(req)->buf */
 		GET_CAP_REQ(req)->buf = buf;
+		buf = NULL;
+	} else {
+		GET_CAP_REQ(req)->buf = NULL;
```
Aren't you missing a `net_buf_unref(buf);` in this branch? Seems like you are, and then you can move the `buf = NULL;` to be a single common step after both branches.

MarkWangChinese replied:
The `buf` comes from the l2cap through the L2CAP `recv` callback `bt_avdtp_l2cap_recv`. avdtp doesn't ref the `buf`, so I think avdtp doesn't need to unref it too.
I change the `req->func(req)` callback function to pass the `buf` as one parameter not give the `buf` to another variable, so I think the `ref/unref` is not needed. please see the last commit in the pr.

jhedberg replied:
Then this function doesn't own the ref, and if you don't own the ref then you can't move it either. In that case the only thing the function needs to do is a `ref()` as part of assigning to the `req`.

MarkWangChinese replied:
I have changed the codes, there is no assigning `buf` to the `req` now. I pass the `buf` as parameter that has similar behavior like the L2CAP callback.

jhedberg commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -321,14 +321,16 @@ static void avdtp_discover_rsp(struct bt_avdtp *session, struct net_buf *buf, ui
+		buf = NULL;
 	} else {
 		DISCOVER_REQ(req)->buf = NULL;
 	}
```
Same here.
